tests/unit/dbapi_test.py

```diff
@@ -0,0 +1,133 @@
+
+    def test_make_config_host_port(self):
+        cfg = _make_config(host="foo.com", port=1234)
+        target = config_with_values(cluster_members=["foo.com:1234"])
+        self.assertEqualConfig(target, cfg)
+
+    def test_make_config_dsn(self):
+        test_cases = [
+            ("hz://", config_with_values(cluster_members=["localhost:5701"])),
+            ("hz://foo.com", config_with_values(cluster_members=["foo.com:5701"])),
+            ("hz://:1234", config_with_values(cluster_members=["localhost:1234"])),
+            ("hz://foo.com:1234", config_with_values(cluster_members=["foo.com:1234"])),
+            (
+                "hz://user:pass@foo.com:1234",
+                config_with_values(
+                    cluster_members=["foo.com:1234"],
+                    creds_username="user",
+                    creds_password="pass",
+                ),
+            ),
+            (
+                "hz://foo.com:1234?cluster.name=prod",
+                config_with_values(
+                    cluster_members=["foo.com:1234"],
+                    cluster_name="prod",
+                ),
+            ),
+            (
+                "hz://foo.com:1234?cluster.name=prod&cloud.token=token1",
+                config_with_values(
+                    cluster_members=["foo.com:1234"],
+                    cluster_name="prod",
+                    cloud_discovery_token="token1",
+                ),
+            ),
+            (
+                "hz://foo.com?smart=false",
+                config_with_values(
+                    cluster_members=["foo.com:5701"],
+                    smart_routing=False,
+                ),
+            ),
+            (
+                "hz://foo.com?ssl=true&ssl.ca.path=ca.pem&ssl.cert.path=cert.pem&ssl.key.path=key.pem&ssl.key.password=123",
+                config_with_values(
+                    cluster_members=["foo.com:5701"],
+                    ssl_enabled=True,
+                    ssl_cafile="ca.pem",
+                    ssl_certfile="cert.pem",
+                    ssl_keyfile="key.pem",
+                    ssl_password="123",
+                ),
+            ),
+        ]
+        for dsn, target in test_cases:
+            cfg = _make_config(dsn=dsn)
+            self.assertEqualConfig(target, cfg)
+
+    def test_make_config_invalid_dsn(self):
+        test_cases = [
+            "http://",
+            "://",
+            "hz://foo.com?smart=False",
+            "hz://foo.com?non.existing=value",
+        ]
+        for dsn in test_cases:
+            self.assertRaises(InterfaceError, lambda: _make_config(dsn=dsn))
+
+    def assertEqualConfig(self, a: Config, b: Config):
+        self.assertEqual(config_to_dict(a), config_to_dict(b))
+
+
+def config_with_values(**kwargs) -> Config:
+    return Config.from_dict(kwargs)
+
+
+def config_to_dict(cfg: Config) -> dict:
+    d = {}
+    for k in cfg.__slots__:
+        d[k] = getattr(cfg, k)
+    return d
```

Comment:
Can you add a test that verifies dsn is parsed correctly when there are problematic characters in it? For example, parsing things like `hz://user?smart=True:pass@foo.com:1234`

Reply:
Which is the problematic character in that example?
We use the standard `urllib.parse.urlparse` function, so anything parsed by that function should be OK.

Reply:
I mean having characters like `:`, `?`, `=` in the values can result in weird parsing if they are not handled correctly